packages/core-plugin-api/src/analytics/AnalyticsDomain.tsx

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { createContext, ReactNode, useContext } from 'react';
+import { AnalyticsDomainValue, AnyAnalyticsDomain } from './types';
+
+export const AnalyticsDomainContext = createContext<AnalyticsDomainValue>({
+  routeRef: 'unknown',
+  pluginId: 'root',
+  componentName: 'App',
+});
+
+/**
+ * A "private" (to this package) hook that enables context inheritance and a
+ * way to read Analytics Domain values at event capture-time.
+ * @private
+ */
+export const useAnalyticsDomain = () => {
+  return useContext(AnalyticsDomainContext);
+};
+
+/**
+ * Provides components in the child react tree an Analytics Domain, ensuring
+ * all analytics events captured within the domain have relevant contextual
+ * attributes.
+ *
+ * Analytics domains are additive, meaning the domain ultimately emitted with
+ * an event is the combination of all domains in the parent tree.
+ */
+export const AnalyticsDomain = ({
+  attributes,
+  children,
+}: {
+  attributes: AnalyticsDomainValue;
+  children: ReactNode;
+}) => {
+  const parentValues = useAnalyticsDomain();
+  const combinedValue = {
+    ...parentValues,
+    ...attributes,
+  };
+
+  return (
+    <AnalyticsDomainContext.Provider value={combinedValue}>
+      {children}
+    </AnalyticsDomainContext.Provider>
+  );
+};
+
+/**
+ * Returns an HOC wrapping the provided component in an Analytics Domain with
+ * the given values.
+ *
+ * @param Component - Component to be wrapped with analytics domain attributes.
+ * @param domain - Analytics domain key/value pairs.
+ */
+export function withAnalyticsDomain<P>(
```

Comment:
Good thoughts!  This ties in pretty closely to that design goal I mentioned earlier where "building Backstage plugins the right way" should get you all the `<AnalyticsDomain />` metadata an app integrator would want for free.  ...And then I realized that not everything actually gets exported from a plugin as a component/react extension (e.g. the `<Link />` component in core components).

I would love to avoid this HOC.  But on the other hand, I'm not sure opening up `useAnalytics()` to domain overrides is the best approach either.  As an example:

```jsx
export const MyThing = ({ myValue }) => {
  const analytics = useAnalytics({
    domain: {
      componentName: 'MyThing',
    },
  });
  const handleClick = () => analytics.captureEvent('click', myValue);
  return <Thing value={myValue} onClick={handleClick} />
};
```

In the above, there's less of a clear separation between information specific to the click event, and metadata about where the click event took place (easy to imagine a plugin developer being confused about what information to put where and why).  Also, if any events are instrumented within `<Thing />`, they'll have no association to `MyThing`, even though they happen within that context.

Another possibility that comes to mind is: avoid this HOC, avoid extending `useAnalytics()`, but fix extensions so that `componentName` works / is meaningful?  Maybe it doesn't even make sense to always force a `domain.componentName = Link` on every link click event, when having the context of the component above it is more helpful anyway.

Reply:
Yeah could be better to have it in the `captureEvent` options as well.

+1 to trying to avoid it altogether though if possible. `componentName` is oddly implementation specific too imo

Reply:
The HOC is dead!